aiohttp/client_reqrep.py

```diff
@@ -428,11 +430,23 @@ def update_transfer_encoding(self) -> None:
             if hdrs.CONTENT_LENGTH not in self.headers:
                 self.headers[hdrs.CONTENT_LENGTH] = str(len(self.body))
 
-    def update_auth(self, auth: Optional[BasicAuth]) -> None:
+    def update_auth(self, auth: Optional[BasicAuth], trust_env: bool = False) -> None:
         """Set basic auth."""
         if auth is None:
             auth = self.auth
         if auth is None:
+            if trust_env:
+                # If no auth is specified, we read from netrc.
+
+                netrc_obj = netrc_from_env()
+                if netrc_obj:
+                    auth_from_netrc = netrc_obj.authenticators(self.url.host)
+                    if auth_from_netrc:
+                        *logins, password = auth_from_netrc
+                        login = logins[0] if logins[0] else logins[-1]
+                        auth = BasicAuth(cast(str, login), cast(str, password))
```

Comment:
Why the cast()? It seems like both of these can be `None`, if `BasicAuth` doesn't allow `None`, then we shouldn't try to pass them here.

Reply:
@Dreamsorcerer ah, I was just copying code from how this is handled in the proxies (in helpers.py), I will take a slightly deeper look onto why the casts are there in the code.

Reply:
Ah, I see, looks like asvetlov was trying to get typing added across the entire codebase, so probably didn't look into it in detail. Probably worth fixing it as it suggests a bug...

If this code comes from there, I also wonder if it's worth deduplicating it.

Reply:
@Dreamsorcerer I've deduplicated it now! I found the type signature for netrc authenticators at https://github.com/python/typeshed/blob/04d4cbd92f2f6b2f845e1c3e0b257a96c3e170a5/stdlib/netrc.pyi#L13, will try to figure out how to type it correctly here shortly

Reply:
@Dreamsorcerer so I fell into a rabbithole, and it actually turns out this behavior changed between 3.10 and 3.11! https://github.com/python/cpython/pull/26330 changed when something returns None vs empty string too.

Reply:
I'll dig in some more and figure out the appropriate thing to do, and leave a lot of comments as well :)

Reply:
Here is some sample code with different netrc files:

```python
import netrc
import tempfile

cases = [
    # Host, with login and password
    "machine testhost login login1 password pass1",
    # Host, with login, account and password
    "machine testhost login login1 account account1 password pass1",
    # Host, with account and password
    "machine testhost account login2 password pass2",
    # Host, with just password, no account nor login
    "machine testhost password pass2",
    # Default with just login, no password
    "default account login3",
]

for line in cases:
    with tempfile.NamedTemporaryFile() as f:
        f.write(line.encode())
        f.flush()
        n = netrc.netrc(f.name)
        print(n.authenticators("testhost"))
```

Here is the output for python3.10:

```
('login1', None, 'pass1')
('login1', 'account1', 'pass1')
('', 'login2', 'pass2')
('', None, 'pass2')
Traceback (most recent call last):
  File "/Users/yuvipanda/code/aiohttp/test_netrc.py", line 21, in <module>
    n = netrc.netrc(f.name)
  File "/opt/homebrew/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/netrc.py", line 31, in __init__
    self._parse(file, fp, default_netrc)
  File "/opt/homebrew/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/netrc.py", line 82, in _parse
    raise NetrcParseError(
netrc.NetrcParseError: malformed default entry default terminated by '' (/var/folders/p4/gnx7h_mj73v92nj_v93475v80000gn/T/tmp8a66t8qr, line 1)
```

Here is the output for 3.11:

```
('login1', '', 'pass1')
('login1', 'account1', 'pass1')
('', 'login2', 'pass2')
('', '', 'pass2')
('', 'login3', '')
```

ignoring the exception, the basic difference seems to be that until 3.10, the type of any of those fields is really `Optional[str]`, while with 3.11 the type is just `str`. Given we want to be compatible, I'm going to treat everything as if it is `Optional[str]`, but set None values to '' to match 3.11 behavior.

Reply:
@Dreamsorcerer ok I've updated the code to be clearer now! Let me know what you think :)

I'll also add test cases based on the helper script I pasted above soon.

Reply:
Typeshed fix: https://github.com/python/typeshed/pull/9376